File: AOC_2023/Day 2/2.cpp
// 12 red cubes, 13 green cubes, and 14 blue cubes

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Iteration{
    int red, green, blue;
};

class Game{
    public:
    vector<Iteration> iterations;
    int minRed = 0, minGreen = 0, minBlue = 0;
    Game(string line){
        string inp = line.substr(line.find(":")+2);
        stringstream ss(inp);
        string token;
        while (getline(ss, token, ';'))
        {
            Iteration i;
            stringstream ss2(token);
            string token2;
            int red = 0, green = 0, blue = 0;
            while (getline(ss2, token2, ','))
            {
                stringstream ss3(token2);
                int num;
                ss3 >> num;
                if(token2.find("red") != string::npos){
                    red = num;
                }else if(token2.find("green") != string::npos){
                    green = num;
                }else if(token2.find("blue") != string::npos){
                    blue = num;
                }
            }
            i.red = red; i.green = green; i.blue = blue;
            iterations.push_back(i);
        }
    }
    bool isGamePlayable(int red, int green, int blue){
        for (int i = 0; i < iterations.size(); i++)
        {
            if(iterations[i].red > red || iterations[i].green > green || iterations[i].blue > blue){
                return false;
            }
        }
        return true;
    }

    int power(){
        int result;
        for (int i = 0; i < iterations.size(); i++)
        {
            minRed = max(minRed, iterations[i].red);
            minGreen = max(minGreen, iterations[i].green);
            minBlue = max(minBlue, iterations[i].blue);
        }
        result = minRed*minGreen*minBlue;
        return result;
    }

};

class Solution
{
    public:
    int red = 12, green = 13, blue = 14;
    vector<Game> games;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            games.push_back(Game(line));
        }
    }
    void part1(){
        int result = 0;
        for (int i = 0; i < games.size(); i++)
        {
            if(games[i].isGamePlayable(red, green, blue)){
                result += (i+1);
            }
        }
        cout<<"Part 1: "<<result<<"\n";
    }
    void part2(){
        int result = 0;
        for (int i = 0; i < games.size(); i++)
        {
            result += games[i].power();
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 2/2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
	"unicode"
)

func initialize(filename string) ([]Game, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var games []Game
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		itrs := strings.Split(scanner.Text()[strings.Index(scanner.Text(), ":")+2:], ";")
		var itrsList []Iterator
		for _, itr := range itrs {
			cubes := strings.Split(itr, ",")
			nums := make([]int, 3)
			for _, cube := range cubes {
				num,err := strconv.Atoi(strings.TrimSpace(strings.TrimRightFunc(cube, unicode.IsLetter)))
				if err != nil {
					return nil, err
				}
				str := strings.TrimLeftFunc(cube, func(r rune) bool {
					return unicode.IsDigit(r) || unicode.IsSpace(r)
				})
				if str == "red" {
					nums[0] = num
				} else if str == "green" {
					nums[1] = num
				}
				if str == "blue" {
					nums[2] = num
				}
			}
			itrsList = append(itrsList, Iterator{Red: nums[0], Green: nums[1], Blue: nums[2]})
		}
		games = append(games, Game{Iterations: itrsList})
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return games, nil
}

type Iterator struct {
	Red int
	Green int
	Blue int
}

type Game struct {
	Iterations []Iterator
}

func part1(input []Game) {
	res := 0
	for i, game := range input {
		b := true
		for _, itr := range game.Iterations {
			if itr.Red > 12 || itr.Green > 13 || itr.Blue > 14 {
				b = false
				break
			}
		}
		if b {
			res += i+1
		}
	}
	fmt.Println("Part 1:", res)
}

func part2(input []Game) {
	res := 0
	for _, game := range input {
		req := make([]int, 3)
		for _, itr := range game.Iterations {
			req[0] = max(req[0], itr.Red)
			req[1] = max(req[1], itr.Green)
			req[2] = max(req[2], itr.Blue)
		}
		res += req[0] * req[1] * req[2]
	}
	fmt.Println("Part 2:", res)
}


func main() {
	filename := "input.txt"
	games, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(games)
	part2(games)
}